mod/state-transition/pkg/core/state_processor_staking.go

```diff
@@ -171,6 +173,115 @@ func (sp *StateProcessor[
 		math.Gwei(sp.cs.MaxEffectiveBalance()),
 	)
 
+	// BeaconKit enforces a cap on the validator set size. If the deposit
+	// breaches the cap, we find the validator with the smallest stake and
+	// mark it as withdrawable so that it will be eventually evicted and
+	// its deposits returned.
+	capHit, err := sp.isValidatorCapHit(st)
+	if err != nil {
+		return err
+	}
+	if !capHit {
+		return sp.addValidatorInternal(st, val, dep.GetAmount())
+	}
+
+	// Adding the validator would breach the cap. Find the validator
+	// with the smallest stake among current and candidate validators
+	// and kit it out.
+	currSmallestVal, err := sp.findSmallestValidator(st)
+	if err != nil {
+		return err
+	}
+
+	slot, err := st.GetSlot()
+	if err != nil {
+		return err
+	}
+	epoch := sp.cs.SlotToEpoch(slot)
+
+	if val.GetEffectiveBalance() <= currSmallestVal.GetEffectiveBalance() {
+		// in case of tie-break among candidate validator we prefer
+		// existing one so we mark candidate as withdrawable
+		val.SetWithdrawableEpoch(epoch)
+		return sp.addValidatorInternal(st, val, dep.GetAmount())
+	}
+
+	// mark exiting validator for eviction and add candidate
+	currSmallestVal.SetWithdrawableEpoch(epoch)
+	idx, err := st.ValidatorIndexByPubkey(currSmallestVal.GetPubkey())
+	if err != nil {
+		return err
+	}
+	if err = st.UpdateValidatorAtIndex(idx, currSmallestVal); err != nil {
+		return err
+	}
+	return sp.addValidatorInternal(st, val, dep.GetAmount())
+}
+
+func (sp *StateProcessor[
+	_, _, _, BeaconStateT, _, _, _, _, _, _, _, _, _, _, _, _, _,
+]) isValidatorCapHit(st BeaconStateT) (bool, error) {
+	if sp.processingGenesis {
+		// minor optimization: we do check in Genesis that
+		// validators do no exceed cap. So hitting cap here
+		// should never happen
+		return false, nil
+	}
+	validators, err := st.GetValidators()
+	if err != nil {
+		return false, err
+	}
+
+	//#nosec:G701 // no overflow risk here
+	return uint32(len(validators)) >= sp.cs.GetValidatorSetCapSize(), nil
+}
+
+// TODO: consider moving this to BeaconState directly
+func (sp *StateProcessor[
+	_, _, _, BeaconStateT, _, DepositT, _, _, _, _, _, _, ValidatorT, _, _, _, _,
+]) findSmallestValidator(st BeaconStateT) (
+	ValidatorT,
+	error,
+) {
+	var smallestVal ValidatorT
+
+	// candidateVal would breach the cap. Find the validator with the
+	// smallest stake among current and candidate validators.
+	currentVals, err := st.GetValidatorsByEffectiveBalance()
```

Comment:
I think this is a bug: I should only for validators in the current active set.
Consider the case:

- validator set is at cap
- a large deposit comes that cause a validator to be mark for ejection
- a second large deposit comes that should cause yet another validator to be ejected.

I think current implementation would pick exactly the first ejected validator as validator to be ejected. So in the end a single validator would be marked for ejection twice and the cap would be breached.

TODO: confirm this with a UT and fix